examples/userapi/05_conditional_dimension.ipynb

```diff
@@ -0,0 +1,858 @@
+      "\n",
+      "    .. code-block:: C\n",
+      "\n",
+      "        for (int i = i_m; i <= i_M; i += 1) {\n",
+      "          if (g[i] > 0 && g[i] < 4) {\n",
+      "            f[g[i]] = f[g[i]] + 1;\n",
+      "          }\n",
+      "        }\n",
+      "\n",
+      "    \n"
+     ]
+    }
+   ],
+   "source": [
+    "from devito import ConditionalDimension\n",
+    "print(ConditionalDimension.__doc__)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Use case A (honour a certain condition)\n",
+    "\n",
+    "\n",
+    "## Defining a condition\n",
+    "Now it's time to show a more descriptive example. We define a conditional that increments by one all values of a Function that are larger than zero. We name our `ConditionalDimension` `ci`.\n",
+    "In this example we want to update again by one all the elements in `f` that are larger than zero, $f(x,y) > 0$\n",
+    "Before updating the elements we initialize our data again to the eye function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Operator `Kernel` run in 0.01 s\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "Data([[2., 1., 1., 1., 1., 1., 1., 1., 1., 1.],\n",
+       "      [1., 2., 1., 1., 1., 1., 1., 1., 1., 1.],\n",
+       "      [1., 1., 2., 1., 1., 1., 1., 1., 1., 1.],\n",
+       "      [1., 1., 1., 2., 1., 1., 1., 1., 1., 1.],\n",
+       "      [1., 1., 1., 1., 2., 1., 1., 1., 1., 1.],\n",
+       "      [1., 1., 1., 1., 1., 2., 1., 1., 1., 1.],\n",
+       "      [1., 1., 1., 1., 1., 1., 2., 1., 1., 1.],\n",
+       "      [1., 1., 1., 1., 1., 1., 1., 2., 1., 1.],\n",
+       "      [1., 1., 1., 1., 1., 1., 1., 1., 2., 1.],\n",
+       "      [1., 1., 1., 1., 1., 1., 1., 1., 1., 2.]], dtype=float32)"
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from devito import Gt\n",
+    "\n",
+    "f.data[:, :] = np.eye(10)\n",
+    "\n",
+    "# Define the condition f(x, y) > 0\n",
+    "condition = Gt(f, 0)\n",
+    "\n",
+    "# Construct a ConditionalDimension\n",
+    "ci = ConditionalDimension(name='ci', parent=y, condition=condition)\n",
+    "\n",
+    "op = Operator(Eq(f, f + 1))\n",
+    "op.apply()\n",
+    "f.data\n",
+    "# print(op.ccode) # Uncomment to view code\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we look at the generated code, we will see that it is identical to that from cell 3 -- no conditionals are generated. This is simply because we construct `ci` but we're not using it. Neither `f` nor `op` know about `ci` yet. We need to pass `ci` as an \"implicit `Dimension`\" to the equation that we want to be conditionally executed as shown in the next cell."
```

Comment:
> If we look at the generated code, we will see that it is identical to that from cell 3 -- no conditionals are generated. This is simply because we construct `ci` but we're not using it. Neither `f` nor `op` know about `ci` yet.

I would drop it and replace with something neater, as

> We've constructed `ci`, but it's still unused, so `op1` is still identical to `op0`.

but now you have to name the Operators differently -- op0, op1, ... -- in increasing order, so you can explicitly refer to them

Reply:
Fixed